src/main/java/edu/harvard/iq/dataverse/api/DownloadInstanceWriter.java

```diff
@@ -80,7 +81,28 @@ public void writeTo(DownloadInstance di, Class<?> clazz, Type type, Annotation[]
             StorageIO<DataFile> storageIO = DataAccess.getStorageIO(dataFile, daReq);
 
             if (storageIO != null) {
+                List<Range> ranges = new ArrayList<>();
+                long fileSize = storageIO.getDataFile().getFilesize();
```

Comment:
So the reason we tried this hack, of getting the size from the underlying DataFile `fileSize` db attribute (`storageIO.getDataFile().getFilesize()`) - that was because `storageIO.getSize()` was returning NULL, correct? 
That was simply because `.open()` hasn't been called on the StorageIO yet.
Note that `storageIO.getSize()` is working further down in this class; the method
`getContentSize()` relies on it. For example, in this line further down (430):
```
if ((contentSize = getContentSize(storageIO)) > 0 && ranges.isEmpty()) {
```
(Note that `getContentSize()` does something extra for tabular files - because of the variable name header that we add in real time)

But it's not just a matter of moving that `StorageIO.open()` line (106) 30 lines up either. I believe it should be the other way around. I.e., we should be checking that Range header and setting the offset much further down in the code (closer to the line 430 quoted above). At which point we have already checked all the extra arguments potentially passed to the Access API, and the method `getContentSize(storageIO)` can be trusted. 

Where we are now, we instantiate the StorageIO and open it for reading, to confirm that this is a valid file that's accessible in storage. But we don't know yet if the stream we have opened is what we are sending to the client - since there are some special cases (in addition to the saved original, there are format conversions and thumbnails, etc.) where we'll need to close it and open another one. 

Sorry for the delay with this. I'll add a bit more. 
This code is a bit of a mess of special cases and it can be confusing. (I have an open issue for refactoring it). But it should still be straightforward-ish to work out the remaining kinks.

Reply:
I'm going to mark this is resolved. Leonid and I have been chatting a fair amount and we're on the same page. I refactored the code to move it down, after the open in 29ea256 and then got offset mostly working again in 4724e11.